src/import.rs

```diff
@@ -1,60 +1,141 @@
     xdcs_path: PathBuf,
     conn: &mut SqliteConnection,
 ) -> anyhow::Result<()> {
-    let dir_entry = std::fs::read_dir(path).context("Failed to read dir")?;
-
-    // this silently skipps files if some file name can not be converted to str
-    let xdcs: Vec<_> = dir_entry
-        .filter_map(|e| e.ok())
-        .map(|e| e.path())
-        .filter(|e| e.is_file())
-        .filter(|file| {
-            file.file_name()
-                .and_then(|a| a.to_str())
-                .map(|a| a.ends_with(".xdc"))
-                .unwrap_or_default()
-        })
-        .collect();
+    let sources_lock =
+        fs::read_to_string(path.join("sources.lock")).context("Failed to read sources.lock")?;
+    let xdc_metas: HashMap<String, WexbdcManifest> = toml::from_str(&sources_lock)?;
 
+    let mut xdcs = vec![];
+    for xdc in xdc_metas.into_values() {
+        let path = PathBuf::from(path).join(&xdc.cache_relname);
+        xdcs.push(tokio::spawn(async move {
+            // compute file size
+            let size = i64::try_from(
+                File::open(&path)
+                    .await
+                    .context("Can't open cache_relname")?
+                    .metadata()
+                    .await?
+                    .len(),
+            )?;
+            // extract icon
+            let reader = ZipFileReader::new(&path).await?;
+            let entries = reader.file().entries();
+            let image = entries
+                .iter()
+                .enumerate()
+                .find(|(_, entry)| {
+                    entry
+                        .entry()
+                        .filename()
+                        .as_str()
+                        .map(|name| name == "icon.png" || name == "icon.jpg")
+                        .unwrap_or_default()
+                })
+                .map(|a| a.0);
+            let image = if let Some(index) = image {
+                let res = read_vec(&reader, index).await?;
+                encode(&res)
+            } else {
+                bail!("Could not find image")
+            };
+
+            Ok(AppInfo {
+                id: 0,
+                app_id: xdc.app_id,
+                version: xdc.version,
+                date: OffsetDateTime::parse(&xdc.date, &Rfc3339)?.unix_timestamp(),
+                name: xdc.name, // xdc.name,
+                submitter_uri: None,
+                source_code_url: xdc.source_code_url,
+                image,
+                description: xdc.description,
+                xdc_blob_path: path,
+                size,
+            })
+        }))
+    }
+
+    let xdcs = join_all(xdcs).await;
     if xdcs.is_empty() {
-        eprintln!("No xdcs to add in {}", path.display());
+        eprintln!("No xdcs from {} added", path.display());
         return Ok(());
     }
 
     let mut added = Vec::new();
     let mut updated = Vec::new();
     let mut ignored = Vec::new();
-    let mut failed = Vec::new();
+    let mut failed = 0;
```

Comment:
errors are printed anyway, so this can just be a number. Keeping errors with the new approach would produce worse code imo.

Reply:
there also is less chance for errors now that the input is provided by xdcget's update/export which already performs a lot of checks.